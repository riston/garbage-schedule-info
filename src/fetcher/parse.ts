import * as cheerio from 'cheerio';
import * as L from 'luxon';

import { pipe } from 'fp-ts/function';
import * as E from 'fp-ts/lib/Either';
import * as D from 'io-ts/Decoder';

const NumberFromString: D.Decoder<unknown, number> = pipe(
  D.string,
  D.parse((s) => {
    const str = s.trim();
    if (str === '') {
      return D.success(0);
    }

    const n = parseFloat(s)
    return isNaN(n) ? D.failure(s, 'NumberFromString') : D.success(n)
  })
);

const DateFromString = pipe(
  D.string,
  D.parse((date) => {
    const parsedDate = L.DateTime.fromFormat(date, 'dd.MM.yyyy', { zone: 'utc'});
    return parsedDate.isValid
      ? D.success(parsedDate)
      : D.failure(date, parsedDate.invalidReason ?? 'Invalid date time received');
  }),
);

export const Schedule = pipe(
  D.struct({
    region: D.string,
    house: D.string,
    frequency_in_days: NumberFromString,
    next_visit_day: DateFromString,
  }),
  D.intersect(
    D.partial({
      day: D.string,
      service_type: D.string,
    }),
  ),
);
export type Schedule = D.TypeOf<typeof Schedule>;

export const ScheduleResult = D.array(Schedule);
export type ScheduleResult = D.TypeOf<typeof ScheduleResult>;

export function parse(content: string): E.Either<D.DecodeError, ScheduleResult> {
  const $ = cheerio.load(content);
  const resultTable$ = $('.tulemus_tabel > tbody tr');

  // First row is skipped
  const rows: unknown[] = [];
  for (let index = 1; index < resultTable$.length; index++) {
    const element$ = resultTable$.get(index);
    const columns$ = $('td', element$);
    if (columns$.length !== 6) {
      continue;
    }

    rows.push({
      house: $(columns$.get(0)).text(),
      region: $(columns$.get(1)).text(),
      frequency_in_days: $(columns$.get(2)).text(),
      days: $(columns$.get(3)).text(),
      service_type: $(columns$.get(4)).text().toLowerCase(),
      next_visit_day: $(columns$.get(5)).text(),
    });
  }

  return ScheduleResult.decode(rows);
}