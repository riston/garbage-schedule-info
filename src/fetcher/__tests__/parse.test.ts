import * as fs from 'fs';
import { join } from 'path';
import nock from 'nock';
import { DateTime } from 'luxon';
import * as E from 'fp-ts/Either';

import { parse } from '../parse';

describe('Parsing content', () => {
  it('hello', () => {
    nock('https://www.ragnsells.ee')
      .get('/klienditugi/graafikud')
      .reply(200, 'Content returned');
  });

  it('should parse the single row success case', () => {
    const content = fs.readFileSync(
      join(__dirname, 'fixture', 'example.success.html'),
      'utf-8',
    );
    expect(parse(content)).toEqual(
      E.right([
        {
          house: 'Mägra',
          frequency_in_days: 84,
          next_visit_day: DateTime.fromISO('2021-07-28T00:00:00.000Z', {
            zone: 'utc',
          }),
          region: 'Laadi küla, Häädemeeste, Pärnu maakond',
          service_type: 'olmeprügi',
        },
      ]),
    );
  });

  it('should parse multiple responses', () => {
    const content = fs.readFileSync(
      join(__dirname, 'fixture', 'multiple.success.html'),
      'utf-8',
    );
    const result = parse(content);

    expect(E.right(result)).toMatchSnapshot();
  });

  it('should parse random response', () => {
    const content = fs.readFileSync(
      join(__dirname, 'fixture', 'random.success.html'),
      'utf-8',
    );
    const result = parse(content);

    expect(E.right(result)).toMatchSnapshot();
  });
});
