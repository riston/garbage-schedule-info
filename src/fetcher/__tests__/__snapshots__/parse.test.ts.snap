// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing content should parse multiple responses 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "_tag": "Right",
    "right": Array [
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 4a",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi 22",
        "next_visit_day": "2021-07-09T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "pakendid",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 21",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi 12",
        "next_visit_day": "2021-07-26T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 19",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 5a",
        "next_visit_day": "2021-08-02T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 7,
        "house": "Telliskivi tn 24",
        "next_visit_day": "2021-07-08T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "biolagunevad jäätmed",
      },
      Object {
        "frequency_in_days": 7,
        "house": "Telliskivi tn 30",
        "next_visit_day": "2021-07-08T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "biolagunevad jäätmed",
      },
      Object {
        "frequency_in_days": 0,
        "house": "Telliskivi tn 24",
        "next_visit_day": "2021-07-08T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi 1",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 14,
        "house": "Telliskivi 11",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi 13",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 7,
        "house": "Telliskivi 14",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 15",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 16",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi 8",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 17",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 18",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 7,
        "house": "Telliskivi 22",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 23",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 27",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 29",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 3",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 31",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi 5",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 6",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 7",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 9",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 2",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi 25",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 0,
        "house": "Telliskivi 30",
        "next_visit_day": "2021-07-07T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 0,
        "house": "Telliskivi tn 20",
        "next_visit_day": "2021-07-08T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 20",
        "next_visit_day": "2021-07-09T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "pakendid",
      },
      Object {
        "frequency_in_days": 28,
        "house": "Telliskivi tn 10",
        "next_visit_day": "2021-07-12T00:00:00.000Z",
        "region": "Pärnu linn, Pärnu, Pärnu maakond",
        "service_type": "olmeprügi",
      },
    ],
  },
}
`;

exports[`Parsing content should parse random response 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "_tag": "Right",
    "right": Array [
      Object {
        "frequency_in_days": 28,
        "house": "Terase tn 8",
        "next_visit_day": "2021-07-19T00:00:00.000Z",
        "region": "Rakvere linn, Lääne-Viru maakond",
        "service_type": "olmeprügi",
      },
    ],
  },
}
`;
